controllers/topology/desired_state.go

```diff
@@ -177,14 +182,76 @@ func computeControlPlane(_ context.Context, s *scope.Scope, infrastructureMachin
+		return nil, errors.Wrap(err, "failed to compute version of control plane")
+	}
+	if err := contract.ControlPlane().Version().Set(controlPlane, version); err != nil {
 		return nil, errors.Wrap(err, "failed to set spec.version in the ControlPlane object")
 	}
 
 	return controlPlane, nil
 }
 
+// computeControlPlaneVersion calculates the version of the desired control plane.
+// The version if calculated using the the state of the current machine deployments, the current control plane
+// and the version defined in the topology.
+func computeControlPlaneVersion(s *scope.Scope) (string, error) {
+	desiredVersion := s.Blueprint.Topology.Version
+	// If we are creating a the control plane object (current control plane is nil), use version from topology.
+	if s.Current.ControlPlane == nil || s.Current.ControlPlane.Object == nil {
+		return desiredVersion, nil
+	}
+
+	// Get the current currentVersion of the control plane.
+	currentVersion, err := contract.ControlPlane().Version().Get(s.Current.ControlPlane.Object)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get the version from control plane spec")
+	}
+
+	// Return early if the version is already equal to the desired version
+	// no further checks required.
+	if *currentVersion == desiredVersion {
+		return *currentVersion, nil
+	}
+
+	// Check if the current control plane is upgrading
+	cpUpgrading, err := contract.ControlPlane().IsUpgrading(s.Current.ControlPlane.Object)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to check if control plane is upgrading")
+	}
+	// If the current control plane is upgrading  (still completing a previous upgrade), then do not pick up the desiredVersion yet.
+	// Return the current version of the control plane. We will pickup the new version after the control
+	// plane is stable.
+	if cpUpgrading {
+		return *currentVersion, nil
+	}
+
+	// If the control plane supports replicas, check if the control plane is in the middle of a scale operation.
+	// If yes, then do not pick up the desiredVersion yet. We will pickup the new version after the control plane is stable.
+	if s.Blueprint.Topology.ControlPlane.Replicas != nil {
+		cpScaling, err := contract.ControlPlane().IsScaling(s.Current.ControlPlane.Object)
+		if err != nil {
+			return "", errors.Wrap(err, "failed to check if the control plane is scaling")
+		}
+		if cpScaling {
+			return *currentVersion, nil
+		}
+	}
+
+	// If the control plane is not upgrading or scaling, we can assume the control plane is stable.
+	// However, we should also check for the MachineDeployments to be stable.
+	// If the MachineDeployments are upgrading (still completing a previous upgrade), then do not pick
+	// up the desiredVersion yet. We will pickup the new version after the MachineDeployments are stable.
+	if s.Current.MachineDeployments.AnyIsUpgrading() {
+		return *currentVersion, nil
+	}
+
+	// Control plane and machine deployments are stable.
+	// Ready to pick up the topology version.
```

Comment:
Great work on these comments :) Makes the upgrade flow really straightforward to read through.